prog/dftb+/lib_dftb/dispdftd4.F90

```diff
@@ -0,0 +1,1289 @@
+    real(dp), allocatable :: zetadq(:, :), zerodq(:, :)
+    real(dp), allocatable :: zetadcn(:, :), zerodcn(:, :)
+    real(dp), allocatable :: dEdq(:), dEdcn(:)
+    real(dp), allocatable :: c6(:, :), dc6dq(:, :), dc6dcn(:, :)
+
+    energies(:) = 0.0_dp
+    gradients(:, :) = 0.0_dp
+    stress(:, :) = 0.0_dp
+
+    nRef = maxval(calc%numberOfReferences(species))
+    allocate(nNeighbour(nAtom))
+    allocate(zetaVec(nRef, nAtom), zeroVec(nRef, nAtom), zetadq(nRef, nAtom),&
+        & zetadcn(nRef, nAtom), zerodcn(nRef, nAtom), zerodq(nRef, nAtom),&
+        & c6(nAtom, nAtom), dc6dq(nAtom, nAtom), dc6dcn(nAtom, nAtom),&
+        & dEdq(nAtom), dEdcn(nAtom))
+
+    dEdq(:) = 0.0_dp
+    dEdcn(:) = 0.0_dp
+
+    call weightReferences(calc, nAtom, species, cn, q, zetaVec, zeroVec, zetadq,&
+        & zetadcn, zerodcn)
+
+    call getAtomicC6(calc, nAtom, species, zetaVec, zetadq, zetadcn,&
+        & c6, dc6dcn, dc6dq)
+
+    call getNrOfNeighboursForAll(nNeighbour, neigh, calc%cutoffInter)
+    do iAt1 = 1, nAtom
+      iSp1 = species(iAt1)
+      do iNeigh = 1, nNeighbour(iAt1)
+        iAt2 = neigh%iNeighbour(iNeigh, iAt1)
+        vec(:) = coords(:, iAt1) - coords(:, iAt2)
+        iAt2f = img2CentCell(iAt2)
+        iSp2 = species(iAt2f)
+        r2 = neigh%neighDist2(iNeigh, iAt1)
+        r1 = sqrt(r2)
+        r4 = r2 * r2
+        r5 = r4 * r1
+        r6 = r2 * r4
+        r8 = r2 * r6
+        r10 = r2 * r8
+
+        rc = 3.0_dp * calc%sqrtZr4r2(iSp1) * calc%sqrtZr4r2(iSp2)
+        rc1 = calc%a1 * sqrt(rc) + calc%a2
+        rc2 = rc1 * rc1
+        rc6 = rc2 * rc2 * rc2
+        rc8 = rc2 * rc6
+        rc10 = rc2 * rc8
+
+        dc6 = c6(iAt1, iAt2f)
+        dc6dcn1 = dc6dcn(iAt1, iAt2f)
+        dc6dcn2 = dc6dcn(iAt2f, iAt1)
+        dc6dq1 = dc6dq(iAt1, iAt2f)
+        dc6dq2 = dc6dq(iAt2f, iAt1)
+
+        f6 = 1.0_dp / (r6 + rc6)
+        f8 = 1.0_dp / (r8 + rc8)
+        f10 = 1.0_dp / (r10 + rc10)
+
+        df6 = -6.0_dp * r5 * f6 * f6
+        df8 = -8.0_dp * r2 * r5 * f8 * f8
+        df10 = -10.0_dp * r4 * r5 * f10 * f10
+
+        dEr = calc%s6 * f6 + calc%s8 * f8 * rc + calc%s10 * rc * rc * 49.0_dp / 40.0_dp * f10
+        dGr = calc%s6 * df6 + calc%s8 * df8 * rc + calc%s10 * rc * rc * 49.0_dp / 40.0_dp * df10
+
+        energies(iAt1) = energies(iAt1) - dEr*dc6/2
+        if (iAt1 /= iAt2f) then
+          energies(iAt2f) = energies(iAt2f) - dEr*dc6/2
+        end if
+
+        grad = -dGr*dc6 * vec / r1
+        gradients(:, iAt1) = gradients(:, iAt1) + grad
+        gradients(:, iAt2f) = gradients(:, iAt2f) - grad
+
+        sigma = spread(grad, 1, 3) * spread(vec, 2, 3)
+        if (iAt1 /= iAt2f) then
+          stress = stress + sigma
+        else
+          stress = stress + 0.5_dp * sigma
+        end if
+
+        dEdcn(iAt1) = dEdcn(iAt1) - dc6dcn1 * dEr
+        dEdcn(iAt2f) = dEdcn(iAt2f) - dc6dcn2 * dEr
+        dEdq(iAt1) = dEdq(iAt1) - dc6dq1 * dEr
+        dEdq(iAt2f) = dEdq(iAt2f) - dc6dq2 * dEr
```

Comment:
Oh, I overlooked this one! It should be guarded by `if (iAt1 /= iAt2f) then` otherwise the gradient is wrong.